Sources/Regex/Parse/Parse.swift

```diff
@@ -2,158 +2,190 @@
 
 Syntactic structure of a regular expression
 
- Regex          -> '' | Alternation
- Alternation    -> Concatenation ('|' Concatenation)*
- Concatenation  -> Quantification Quantification*
- Quantification -> (Group | Atom) Quantifier?
- Atom           -> <token: .character> | <any> | ... character classes ...
- CaptureGroup   -> '(' RE ')'
- Group          -> '(' '?' ':' RE ')'
-
- Quantifier -> <provided by lexer>
- Group      -> GroupStart Regex ')'
- GroupStart -> <provided by lexer>
+ Regex           -> Alternation
+ Alternation     -> Concatenation ('|' Concatenation)*
+ Concatenation   -> (!'|' !')' ConcatComponent)*
+ ConcatComponent -> Trivia | Quote | Quantification
+ Quantification  -> QuantOperand Quantifier?
+ QuantOperand    -> Group | CustomCharClass | Atom
+ Group           -> GroupStart Regex ')'
+
+Custom character classes are a mini-language to their own. We
+support UTS#18 set operators and nested character classes. The
+meaning of some atoms, such as `\b` changes inside a custom
+chararacter class. Below, we have a grammar "scope", that is we say
+"SetOp" to mean "CustomCharactetClass.SetOp", so we don't have to
+abbreviate/obfuscate/disambiguate with ugly names like "CCCSetOp".
+
+TODO: Hamish, can you look this over?
+Also, PCRE lets you end in `&&`, but not Oniguruma as it's a set
+operator. We probably want a rule similar to how you can end in `-`
+and that's just the character. Perhaps we also have syntax options
+in case we need a compatibilty mode (it's easy to add here and now)
+
+ CustomCharClass -> Start Set (SetOp Set)* ']'
+ Set             -> Member+
+ Member          -> CustomCharClass | !']' !SetOp (Range | Atom)
+ Range           -> Atom `-` Atom
+
+Lexical analysis provides the following:
+
+ Atom       -> `lexAtom`
+ Trivia     -> `lexComment` | `lexNonSemanticWhitespace`
+ Quote      -> `lexQuote`
+ Quantifier -> `lexQuantifier`
+ GroupStart -> `lexGroupStart`
+
+ CustomCharacterClass.Start -> `lexCustomCCStart`
+ CustomCharacterClass.SetOp -> `lexCustomCCBinOp`
+
+ better design for "metacharacter"
+
+
+ Escaped -> UniScalar | TODO
+ UniScalar -> 'u{' HexDigit{1, 8}
+            | 'u' HexDigit{4}
+            | 'x{' HexDigit{1, 8}
+            | 'x' HexDigit{2}
+            | 'U' HexDigit{8}
+
+
+ // TODO: Hamish, can you look over this?
+ CustomCharClass -> '[' '^'? CCCComponent+ ']'
+ CCCComponent    -> CharSet SetOp CharSet | Atom
+ CCCBinOp        ->
```

Comment:
Is this meant to be part of the `CustomCharClass` definition above?

Reply:
Whoops, I think I just ended up doing it instead 😅. That was an older version of `CustomCharClass` above.